File: solutions/java/src/snakeandladdergame/SnakeAndLadderGame.java
package snakeandladdergame;

import java.util.ArrayList;
import java.util.List;

public class SnakeAndLadderGame {
    private final Board board;
    private final List<Player> players;
    private final Dice dice;
    private int currentPlayerIndex;

    public SnakeAndLadderGame(Board board, List<String> playerNames, Dice dice) {
        this.board = board;
        this.dice = dice;
        this.players = new ArrayList<>();
        for (String playerName : playerNames) {
            players.add(new Player(playerName));
        }
        currentPlayerIndex = 0;
    }

    public void play() {
        while (!isGameOver()) {
            Player currentPlayer = players.get(currentPlayerIndex);
            int diceRoll = dice.roll();
            int newPosition = currentPlayer.getPosition() + diceRoll;

            if (newPosition <= board.getBoardSize()) {
                currentPlayer.setPosition(board.getNextPosition(newPosition));
                System.out.println(currentPlayer.getName() + " rolled a " + diceRoll +
                        " and moved to position " + currentPlayer.getPosition());
            }

            if (currentPlayer.getPosition() == board.getBoardSize()) {
                System.out.println(currentPlayer.getName() + " wins!");
                break;
            }

            currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        }
    }

    private boolean isGameOver() {
        for (Player player : players) {
            if (player.getPosition() == board.getBoardSize()) {
                return true;
            }
        }
        return false;
    }
}

File: solutions/golang/snakeandladdergame/snake_and_ladder_game.go
package snakeandladdergame

import (
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

type SnakeAndLadderGame struct {
	ID               int64
	Board            *Board
	Players          []*Player
	Dice             *Dice
	CurrentPlayerIdx int
}

func NewSnakeAndLadderGame(playerNames []string) *SnakeAndLadderGame {
	game := &SnakeAndLadderGame{
		ID:               generateID(),
		Board:            NewBoard(),
		Dice:             NewDice(),
		Players:          []*Player{},
		CurrentPlayerIdx: 0,
	}

	for _, name := range playerNames {
		game.Players = append(game.Players, NewPlayer(name))
	}
	return game
}

func (g *SnakeAndLadderGame) Play(wg *sync.WaitGroup) {
	for !g.isGameOver() {
		player := g.Players[g.CurrentPlayerIdx]
		roll := g.Dice.Roll()
		newPosition := player.Position + roll

		if newPosition <= g.Board.Size {
			player.Position = g.Board.GetNewPosition(newPosition)
			fmt.Printf("Game: %d :- %s rolled a %d and moved to position %d\n", g.ID, player.Name, roll, player.Position)
		}

		if player.Position == g.Board.Size {
			fmt.Printf("For Game %d :- %s wins!\n", g.ID, player.Name)
			break
		}

		g.CurrentPlayerIdx = (g.CurrentPlayerIdx + 1) % len(g.Players)
		time.Sleep(10 * time.Millisecond) //To simulate the demo
	}
	wg.Done()
}

func (g *SnakeAndLadderGame) isGameOver() bool {
	for _, player := range g.Players {
		if player.Position == g.Board.Size {
			return true
		}
	}
	return false
}

var idCounter int64

func generateID() int64 {
	return atomic.AddInt64(&idCounter, 1)
}
